File: hooks/stop-prompt-architecture-md.py
#!/usr/bin/env python3
"""Stop hook: prompt to create or update ARCHITECTURE.md after file edits."""

import json
import sys
from pathlib import Path

EDIT_TOOLS = {"Edit", "Write"}
ARCHITECTURE_FILENAME = "ARCHITECTURE.md"


def _scan_transcript(transcript_path: str) -> tuple[bool, bool]:
    """Return (edited_non_architecture, edited_architecture) from transcript."""
    edited_non_architecture = False
    edited_architecture = False

    try:
        with open(transcript_path, encoding="utf-8") as f:
            lines = f.readlines()
    except OSError:
        return False, False

    for line in lines:
        try:
            entry = json.loads(line)
        except json.JSONDecodeError:
            continue

        content = entry.get("message", {}).get("content", [])
        if not isinstance(content, list):
            continue

        for block in content:
            if not isinstance(block, dict):
                continue
            if block.get("type") != "tool_use":
                continue
            if block.get("name") not in EDIT_TOOLS:
                continue

            file_path: str = block.get("input", {}).get("file_path", "")
            if not file_path:
                continue

            if Path(file_path).name == ARCHITECTURE_FILENAME:
                edited_architecture = True
            else:
                edited_non_architecture = True

    return edited_non_architecture, edited_architecture


def main() -> None:
    data = json.load(sys.stdin)

    if data.get("stop_hook_active"):
        return

    if data.get("permission_mode") == "plan":
        return

    transcript_path = data.get("transcript_path", "")
    if not transcript_path:
        return

    edited_files, edited_arch = _scan_transcript(transcript_path)

    if not edited_files:
        return

    cwd = data.get("cwd", "")
    arch_exists = (Path(cwd) / ARCHITECTURE_FILENAME).is_file() if cwd else False

    if not arch_exists:
        json.dump(
            {
                "decision": "block",
                "reason": (
                    "ARCHITECTURE.md が存在しません。\n"
                    "ファイル構成の変更を反映するため作成してください。"
                ),
            },
            sys.stdout,
        )
        return

    if not edited_arch:
        json.dump(
            {
                "decision": "block",
                "reason": (
                    "ファイルを編集しましたが ARCHITECTURE.md を更新していません。\n"
                    "変更内容を反映してください。"
                ),
            },
            sys.stdout,
        )


if __name__ == "__main__":
    main()

File: hooks-rs/src/hooks/stop_prompt_architecture_md.rs
use std::path::Path;

use crate::input::HookInput;
use crate::output::block;
use crate::transcript;

const ARCHITECTURE_FILENAME: &str = "ARCHITECTURE.md";
const EDIT_TOOLS: &[&str] = &["Edit", "Write"];

pub fn run(input: &HookInput) {
    if input.stop_hook_active {
        return;
    }
    if input.permission_mode == "plan" {
        return;
    }

    let transcript_path = &input.transcript_path;
    if transcript_path.is_empty() {
        return;
    }

    let entries = transcript::read_transcript(Path::new(transcript_path));
    let tool_uses = transcript::extract_tool_uses(&entries);

    let mut edited_non_arch = false;
    let mut edited_arch = false;

    for tu in &tool_uses {
        if !EDIT_TOOLS.contains(&tu.tool_name.as_str()) {
            continue;
        }
        if Path::new(&tu.file_path)
            .file_name()
            .and_then(|n| n.to_str())
            == Some(ARCHITECTURE_FILENAME)
        {
            edited_arch = true;
        } else {
            edited_non_arch = true;
        }
    }

    if !edited_non_arch {
        return;
    }

    let cwd = &input.cwd;
    let arch_exists = if !cwd.is_empty() {
        Path::new(cwd).join(ARCHITECTURE_FILENAME).is_file()
    } else {
        false
    };

    if !arch_exists {
        block(
            "ARCHITECTURE.md が存在しません。\nファイル構成の変更を反映するため作成してください。",
        );
        return;
    }

    if !edited_arch {
        block(
            "ファイルを編集しましたが ARCHITECTURE.md を更新していません。\n変更内容を反映してください。",
        );
    }
}
